src/dowel/logger.py

```diff
@@ -267,15 +267,21 @@ def dump_output_type(self, output_type, step=None):
         """
         for output in self._outputs:
             if isinstance(output, output_type):
-                output.dump(step=step)
+                try:
+                    output.dump(step=step)
+                except LoggerWarning as e:
+                    self._warn(e.to_string())
 
     def dump_all(self, step=None):
         """Dump all outputs connected to the logger.
 
         :param step: The current run step.
         """
         for output in self._outputs:
-            output.dump(step=step)
+            try:
+                output.dump(step=step)
+            except LoggerWarning as e:
```

Comment:
why not let the output just print the warning itself? Warnings are not exceptions -- they don't break flow control.

Reply:
Three reasons,
- An output will need a `self._warn()` to print the warning itself, either by making its own, or borrowing from the `Logger`. If it makes its own, the `Logger` loses control of warning printing (and there will be duplicate code). If it borrows from the `Logger`, it will require the `Logger` instance passed to the `Output`. 
- By raising a warning exception, an output asks the logger to handle the printing at its discretion. 
- Warnings are subclasses of exception. https://docs.python.org/3/library/warnings.html#warning-categories. And I see a warning class defined in the logger. https://github.com/rlworkgroup/dowel/blob/621d3e421f9553dde474d6d6ba82c91ff2da6821/src/dowel/logger.py#L328 I think this suggests material "sub"-warnings in outputs.

Reply:
These are excellent points, but don't change the fact using exceptions for flow control is a big problem. Furthermore, it's simply not idiomatic python to use `raise` with warnings. That's why the `warnings` module exists.

When they do happen, it's likely that this warnings will be raised repeatedly. Exceptions break most optimizations in interpreters and modern CPUs. Firing a bunch of them repeatedly for flow control can silently hose performance.

`self_warn()` exists so we can only print warnings once. I think you have run afoul of this TODO: https://github.com/rlworkgroup/dowel/issues/20.

I suggest you either make a process-global `warn_once` or just spam the user with warnings.

Reply:
I don't feel like fixing #20 here - that will extend the scope of this PR. I will follow `CsvOutput`'s pattern to implement a `self._warn()` in `TensorboardOutput`.